src/dividat-driver/flex/main.go

```diff
@@ -0,0 +1,262 @@
+
+		handle.cancelCurrentConnection = cancel
+	}
+}
+
+// Deregister subscribers and disconnect when none left
+func (handle *Handle) DeregisterSubscriber() {
+	handle.subscriberCount--
+
+	if handle.subscriberCount == 0 && handle.cancelCurrentConnection != nil {
+		handle.cancelCurrentConnection()
+		handle.cancelCurrentConnection = nil
+	}
+}
+
+// Keep looking for serial devices and connect to them when found, sending signals into the
+// callback.
+func listeningLoop(ctx context.Context, logger *logrus.Entry, onReceive func([]byte)) {
+	for {
+		scanAndConnectSerial(ctx, logger, onReceive)
+
+		// Terminate if we were cancelled
+		if ctx.Err() != nil {
+			return
+		}
+
+		time.Sleep(2 * time.Second)
+	}
+}
+
+// One pass of browsing for serial devices and trying to connect to them turn by turn, first
+// successful connection wins.
+func scanAndConnectSerial(ctx context.Context, logger *logrus.Entry, onReceive func([]byte)) {
+	ports, err := enumerator.GetDetailedPortsList()
+	if err != nil {
+		logger.WithField("error", err).Info("Could not list serial devices.")
+		return
+	}
+
+	for _, port := range ports {
+		// Terminate if we have been cancelled
+		if ctx.Err() != nil {
+			return
+		}
+
+		logger.WithField("name", port.Name).WithField("vendor", port.VID).Debug("Considering serial port.")
+
+		if isFlexLike(port) {
+			connectSerial(ctx, logger, port.Name, onReceive)
+		}
+	}
+}
+
+// Check whether a port looks like a potential Flex device.
+//
+// Vendor IDs:
+//   16C0 - Van Ooijen Technische Informatica (Teensy)
+func isFlexLike(port *enumerator.PortDetails) bool {
+	vendorId := strings.ToUpper(port.VID)
+
+	return vendorId == "16C0"
+}
+
+
+// Serial communication
+
+type ReaderState int
+
+const (
+	WAITING_FOR_HEADER ReaderState = iota
+	HEADER_START
+	HEADER_READ_LENGTH_MSB
+	WAITING_FOR_BODY
+	BODY_START
+	BODY_POINT
+	UNEXPECTED_BYTE
+)
+
+const (
+	HEADER_START_MARKER = 'N'
+	BODY_START_MARKER = 'P'
+)
+
+// Actually attempt to connect to an individual serial port and pipe its signal into the callback, summarizing
+// package units into a buffer.
+func connectSerial(ctx context.Context, logger *logrus.Entry, serialName string, onReceive func([]byte)) {
+	mode := &serial.Mode{
+		BaudRate: 115200,
+		Parity:   serial.NoParity,
+		DataBits: 8,
+		StopBits: serial.OneStopBit,
+	}
+
+	START_MEASUREMENT_CMD := []byte{'S', '\n'}
+
+	logger.WithField("name", serialName).Info("Attempting to connect with serial port.")
+	port, err := serial.Open(serialName, mode)
+	if err != nil {
+		logger.WithField("config", mode).WithField("error", err).Info("Failed to open connection to serial port.")
+		return
+	}
+	defer port.Close()
+
+	_, err = port.Write(START_MEASUREMENT_CMD)
+	if err != nil {
+		logger.WithField("error", err).Info("Failed to write start message to serial port.")
+		port.Close()
+		return
+	}
+
+	reader := bufio.NewReader(port)
+	state := WAITING_FOR_HEADER
+	pointsLeftInSet := 0
+	bytesLeftInPoint := 0
```

Comment:
What are points and sets referred in these namings ? Are points individual captors in the flex mat ?

Reply:
Renamed "point" to "sample": 476cd27

"Set" might then be clear enough? The device sends sets of samples from the sensors, measured at ~ the same moment in time.

Reply:
:+1: